### PR TITLE
feat(grouping): Replace quoted string values and datetime strings

### PR DESCRIPTION
**Pull request by armenzg**
Some errors can include key/value pairs like `key='some_str'` and `key=datetime.datetime(2023, 5, 2, 23, 25, tzinfo=datetime.timezone.utc)`.

When there's no stacktrace, it produces the errors not to be grouped together.

This change replaces quoted strings for key/value pairs with `<quoted_str>` and full date times with `<date>`.
This will help all those errors to be grouped together while still preserving the values in the title.

See #49681 for details.

### CODE REVIEW

File: tests/sentry/grouping/categorization_inputs/message_with_key_pair_values.json
Comment (jjbayer) [inline tests/sentry/grouping/categorization_inputs/message_with_key_pair_values.json:1]: I think the right place for this is `grouping_inputs` and then running `test_variants.py`:

https://github.com/getsentry/sentry/blob/f025dcf1957323c35458da9584c4a7ecd803771a/tests/sentry/grouping/test_categorization.py#L6-L7

File: tests/sentry/grouping/grouping_inputs/message_with_key_pair_values.json
Comment (armenzg) [inline tests/sentry/grouping/grouping_inputs/message_with_key_pair_values.json:1]: Nothing to see here.

File: tests/sentry/grouping/similarity/snapshots/test_features/test_similarity_extract_grouping_input/message_with_key_pair_values.pysnap.new
Comment (armenzg) [inline tests/sentry/grouping/similarity/snapshots/test_features/test_similarity_extract_grouping_input/message_with_key_pair_values.pysnap.new:1]: Nothing to see here.

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/legacy@2019_03_12/message_with_key_pair_values.pysnap.new
Comment (armenzg) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/legacy@2019_03_12/message_with_key_pair_values.pysnap.new:1]: Nothing to see here.

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/mobile@2021_02_12/message_with_key_pair_values.pysnap.new [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+created: '2023-06-06T13:20:21.816240Z'
+creator: sentry
+source: tests/sentry/grouping/test_variants.py
+---
+default:
+  hash: "e02854e73673e108fe54768e0f19317b"
+  component:
+    default*
+      message* (stripped common values)
+        "Error key1=<quoted_str> key2=<int> key3=False key4=<date> other_date=datetime.datetime(<int>, <int>, <int>, <int>, <int>, tzinfo=d..."
```
Comment (armenzg) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/mobile@2021_02_12/message_with_key_pair_values.pysnap.new:11]: As expected.

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_04_17/message_with_key_pair_values.pysnap.new [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+created: '2023-06-06T13:20:21.867984Z'
+creator: sentry
+source: tests/sentry/grouping/test_variants.py
+---
+default:
+  hash: "e02854e73673e108fe54768e0f19317b"
+  component:
+    default*
+      message* (stripped common values)
+        "Error key1=<quoted_str> key2=<int> key3=False key4=<date> other_date=datetime.datetime(<int>, <int>, <int>, <int>, <int>, tzinfo=d..."
```
Comment (armenzg) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_04_17/message_with_key_pair_values.pysnap.new:11]: As expected.

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_05_08/message_with_key_pair_values.pysnap.new [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+created: '2023-06-06T13:20:21.779482Z'
+creator: sentry
+source: tests/sentry/grouping/test_variants.py
+---
+default:
+  hash: "e02854e73673e108fe54768e0f19317b"
+  component:
+    default*
+      message* (stripped common values)
+        "Error key1=<quoted_str> key2=<int> key3=False key4=<date> other_date=datetime.datetime(<int>, <int>, <int>, <int>, <int>, tzinfo=d..."
```
Comment (armenzg) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_05_08/message_with_key_pair_values.pysnap.new:11]: As expected.

File: tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_10_29/message_with_key_pair_values.pysnap.new [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+created: '2023-06-06T13:20:21.797748Z'
+creator: sentry
+source: tests/sentry/grouping/test_variants.py
+---
+default:
+  hash: "e02854e73673e108fe54768e0f19317b"
+  component:
+    default*
+      message* (stripped common values)
+        "Error key1=<quoted_str> key2=<int> key3=False key4=<date> other_date=datetime.datetime(<int>, <int>, <int>, <int>, <int>, tzinfo=d..."
```
Comment (armenzg) [inline tests/sentry/grouping/snapshots/test_variants/test_event_hash_variant/newstyle@2019_10_29/message_with_key_pair_values.pysnap.new:11]: As expected.

**Review by armenzg — COMMENTED** [review]
This change affects the following grouping configs:
* newstyle (2019-*, 2023-01)
* mobile (2021)
* similarity (2020-07)

At the moment of this PR, 2019-10 is the current grouping config.

**Review by armenzg — COMMENTED** [review]
I've made this change more strict to only change values for key/value pairs and prevent affecting a lot of group messages.

File: src/sentry/grouping/strategies/message.py [hunk 5]
```diff
@@ -81,7 +81,8 @@
             (2[0-3]|[0-1][\d]):([0-5][\d])
             (?::(60|[0-5][\d]))?\s+
             ([-\+][\d]{2}[0-5][\d]|(?:UT|GMT|(?:E|C|M|P)(?:ST|DT)|[A-IK-Z]))
-        )
+        ) |
+        (datetime.datetime\(.*?\))
```
[the remark this answers is not in the document]
Comment (untitaker) [reply src/sentry/grouping/strategies/message.py:85]: *nit*: the first `.` in your regex will match any character, so `datetimeXdatetime` is matched as well

**Review by untitaker — APPROVED** [review]

**Review by Swatinem — APPROVED** [review]
lgtm, though I have ideas for further improvements:

- as Markus suggested, this should be extended to double-quoted strings
- this looks very specific to python and how it serializes its types?
- ☝🏻 related to that, maybe the logger was used in a wrong way? if possible, the parameterized message should be used I guess?
- maybe also extend this to `True`, `False`?

**Comment by untitaker** [comment]
Arpad is actually right about message parameters (but in the end we do not control whether users use logging right). I would recommend adding a testcase where exception value contains python parameters like that, where we don't have a better story.

File: src/sentry/grouping/strategies/message.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (armenzg) [reply src/sentry/grouping/strategies/message.py:85]: Good eye!

**Pull request merged**